server/v2/core/store/types.go

```diff
@@ -6,27 +6,55 @@ import (
 
 type Hash = []byte
 
-var _ store.KVStore = (WritableState)(nil)
+var _ store.KVStore = (Writer)(nil)
 
 // Store defines the underlying storage engine of an app.
 type Store interface {
 	// StateLatest returns a readonly view over the latest
 	// committed state of the store. Alongside the version
 	// associated with it.
-	StateLatest() (uint64, ReadonlyState, error)
+	StateLatest() (uint64, GetReader, error)
 
 	// StateAt returns a readonly view over the provided
 	// state. Must error when the version does not exist.
-	StateAt(version uint64) (ReadonlyState, error)
+	StateAt(version uint64) (GetReader, error)
 
 	// StateCommit commits the provided changeset and returns
 	// the new state root of the state.
-	StateCommit(changes []ChangeSet) (Hash, error)
+	StateCommit(changes []AccountStateChanges) (Hash, error)
 }
 
-// ChangeSet represents a change in a key and value of state.
+// GetReader represents a readonly view over all the accounts state.
+type GetReader interface {
+	// GetAccountReadonlyState must return the state for the provided address.
+	// Storage implements might treat this as a prefix store over an address.
+	// Prefix safety is on the implementer.
+	GetAccountReader(address []byte) (Reader, error)
+}
+
+// WritableAccountsState represents a writable account state.
+type GetWriter interface {
+	GetReader
+	// GetAccountWritableState must the return a WritableState
+	// for the provided account address.
+	GetAccountWriter(address []byte) (Writer, error)
+	// ApplyAccountsStateChanges applies all the state changes
+	// of the accounts. Ordering of the returned state changes
+	// is an implementation detail and must not be assumed.
+	ApplyStateChanges(stateChanges []AccountStateChanges) error
+	// GetAccountsStateChanges returns the list of the state
+	// changes so far applied. Order must not be assumed.
+	GetStateChanges() ([]AccountStateChanges, error)
+}
+
+type AccountStateChanges struct {
```

Comment:
I think we should find a better name for `AccountStateChanges` and `StateChange` :thinking-face:

Reply:
did statchanges and kvpair, follows closer to store